### ISSUE TITLE
Library fails to get URL the second or third time.

### ISSUE THREAD

**Issue by timothydillan** [horihiro/esp8266-google-tts#10]
Hi there.

First of all, I'd like to say thanks for your library.

I have a problem, that is, I'm only able to get the TTS link only once in a loop.

Once, the next loop comes in, the TTS library either takes a very long time to generate a link, or, just doesn't generate a link at all.

I've also edited the original library, so that it uses setInsecure instead of setFingerprint.

Only thing I've found is that the WiFi client is stuck at 0 on the next iteration. (```while (client.available == 0)```)

Here's my code:
```
#include <ESP8266WiFi.h>
#include <ArduinoJson.h>
#include <SPI.h>
#include <RFID.h>
#include <Wire.h> 
#include <LiquidCrystal_I2C.h>
#include "Tone.h"
#include <google-tts.h>
#include "AudioFileSourceICYStream.h"
#include "AudioFileSourceBuffer.h"
#include "AudioGeneratorMP3.h"
#include "AudioOutputI2SNoDAC.h"

TTS tts;

LiquidCrystal_I2C lcd(0x27, 16, 2);

#define SS_PIN D3
#define RST_PIN D4
#define Buzzer D8

const char* wifiName = "TD Samsung";
const char* wifiPass = "*******";

unsigned long currentMillis;

RFID rfid(SS_PIN, RST_PIN);

void MDCallback(void *cbData, const char *type, bool isUnicode, const char *string)
{
  const char *ptr = reinterpret_cast<const char *>(cbData);
  (void) isUnicode; 
  char s1[32], s2[64];
  strncpy_P(s1, type, sizeof(s1));
  s1[sizeof(s1)-1]=0;
  strncpy_P(s2, string, sizeof(s2));
  s2[sizeof(s2)-1]=0;
  Serial.printf("METADATA(%s) '%s' = '%s'\n", ptr, s1, s2);
  Serial.flush();
}

void StatusCallback(void *cbData, int code, const char *string)
{
  const char *ptr = reinterpret_cast<const char *>(cbData);
  char s1[64];
  strncpy_P(s1, string, sizeof(s1));
  s1[sizeof(s1)-1]=0;
  Serial.printf("STATUS(%s) '%d' = '%s'\n", ptr, code, s1);
  Serial.flush();
}

void setup() {
  Serial.begin(115200);
  SPI.begin();
  rfid.init();
  delay(10);
  pinMode(Buzzer, OUTPUT);

  Wire.begin(D2, D1);
  
  lcd.begin();
  
  lcd.backlight();
  
  Serial.println();
  
  Serial.print("Connecting to ");
  Serial.println(wifiName);

  lcd.setCursor(1,0);
  lcd.print("Initialization");
  lcd.setCursor(2,1);
  lcd.print("Connecting..."); 

  WiFi.disconnect();
  WiFi.mode(WIFI_STA);
  WiFi.begin(wifiName, wifiPass);

  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }

  Serial.println("");
  Serial.println("WiFi connected");
  Serial.println("IP address: ");
  Serial.println(WiFi.localIP());   //You can get IP address assigned to ESP

  ModeDevice();
}

void ModeDevice()
{
  HTTPClient https;

  Serial.print("Request Link:");
  Serial.println(hostMode);
  
  https.begin(hostMode);
  
  int httpCode = https.GET();            //Send the request
  String payload = https.getString();    //Get the response payload from server

  Serial.print("Response Code:"); //200 is OK
  Serial.println(httpCode);       //Print HTTP return code

  Serial.print("Returned data from Server:");
  Serial.println(payload);    //Print request response payload
  
  if (httpCode == 200)
  {
    DynamicJsonDocument doc(1024);
  
   // Parse JSON object
    auto error = deserializeJson(doc, payload);
    if (error) {
      Serial.print(F("deserializeJson() failed with code "));
      Serial.println(error.c_str());
      return;
    }
  
    // Decode JSON/Extract values
    String responStatus = doc["status"].as<String>();
    String responMode = doc["mode"].as<String>();
    String responKet = doc["ket"].as<String>();

    Serial.println();
    Serial.print("status : ");
    Serial.println(responStatus);
    
    Serial.print("mode : ");
    Serial.println(responMode);
    
    Serial.print("ket : ");
    Serial.println(responKet);
    Serial.println("-------------------");
    Serial.println();

    lcd.clear();
    lcd.setCursor(0, 0);
    lcd.print("Dismissal System");
    
    if (responMode == "SCAN"){
      ModeAlat = "SCAN";
      lcd.setCursor(0,1);
      lcd.print("Scan your card>>");
    }else{
      ModeAlat = "";
      lcd.setCursor(0,1);
      lcd.print(responKet);
    }
    
  } else { 
    lcd.clear();
    lcd.setCursor(0,0);
    lcd.print("Connection error.");
    Serial.println();
    Serial.println("Koneksi Error");
    lcd.print("                   ");
  }

  Serial.println("Closing connection.");
  Serial.println();
  Serial.println("Scan a card.");

  https.end();

  delay(100);
}

void speak(const char *url) {
  
  AudioGeneratorMP3 *mp3;
  AudioFileSourceICYStream *file;
  AudioFileSourceBuffer *buff;
  AudioOutputI2SNoDAC *out;
  
  bool audioPlaying = true;
  bool hasSetUpAudio = false;
  
  Serial.println("Start of TTS function.");
  // While there are no audio playing,
  while (audioPlaying) {
    // and if audio hasn't been set up'ed,
    if (!hasSetUpAudio) {
      audioLogger = &Serial;
      file = new AudioFileSourceICYStream(url);
      file->RegisterMetadataCB(MDCallback, (void*)"ICY");
      buff = new AudioFileSourceBuffer(file, 4096);
      buff->RegisterStatusCB(StatusCallback, (void*)"buffer");
      out = new AudioOutputI2SNoDAC();
      mp3 = new AudioGeneratorMP3();
      mp3->RegisterStatusCB(StatusCallback, (void*)"mp3");
      Serial.println("MP3 begin.");
      mp3->begin(buff, out);
      // and set hasSetUpAudio to true, so that this code won't be executed twice.
      hasSetUpAudio = true;
    }

    static int lastms = 0;
    
    // While the audio is playing,
    if (mp3->isRunning()) {
      //Serial.println("MP3 is running.");
      // Check if we've played for a second,
      if (millis() - lastms >= 1000) {
        lastms = millis();
        // and show that we've played for a second.
        Serial.printf("Running for %d ms...\n", lastms);
       }
      // If the mp3 is not looping,
      if (!mp3->loop()) {
        // stop the MP3.
        mp3->stop();
      }
    } else {
      // If the audio is not playing, print it and
      // set audioplaying to true.
      Serial.println("MP3 done");
      audioPlaying = false;
    }
  }
  delete mp3, file, buff, out;
  mp3 = NULL;
  file = NULL;
  buff = NULL;
  out = NULL;
  Serial.println("End of TTS function.");
}

void loop() {
    
  if (ModeAlat != "SCAN")
    return;

  rfid.halt();
    
  if (!rfid.isCard() || !rfid.readCardSerial()) {
    lcd.setCursor(0, 0);
    lcd.print("Dismissal System");
    lcd.setCursor(0,1);
    lcd.print("Scan your card>>");
    return;
  }

  currentMillis = millis();

  Serial.println("Card found");

  String RFID = String(rfid.serNum[0],HEX) +"-"+ String(rfid.serNum[1],HEX) +"-"+ String(rfid.serNum[2],HEX) +"-"+ String(rfid.serNum[3],HEX) +"-"+ String(rfid.serNum[4],HEX);

  tone(Buzzer, NOTE_E7, 100);

  Serial.print("RFID: ");
  Serial.println(RFID);
  Serial.println("");

  String host = hostSCAN;
  host += "&rfid=";
  host += RFID;
  HTTPClient https;
  Serial.print("Request Link:");
  Serial.println(host);
  Serial.println("");
  
  https.begin(host);
  
  int httpCode = https.GET();            //Send the GET request
  String payload = https.getString();    //Get the response payload from server

  https.end();

  Serial.print("Response Code:"); //200 is OK
  Serial.println(httpCode);       //Print HTTP return code
  Serial.println("");

  Serial.print("Returned data from Server: ");
  Serial.println(payload);    //Print request response payload
  Serial.println("");
  
  if (httpCode != 200) {
    lcd.setCursor(0, 0);
    lcd.print("Dismissal System");
    lcd.setCursor(0,1);
    lcd.print("HTTP: " + httpCode);
    return;
  }
    
  DynamicJsonDocument doc(1024);

  auto error = deserializeJson(doc, payload);
  
  if (error) {
    Serial.print(F("deserializeJson() failed with code "));
    Serial.println(error.c_str());
    return;
  }

  String responStatus = doc["status"].as<String>();
  String responKeterangan = doc["ket"].as<String>();
  String responNama = doc["nama"].as<String>();
  String responLanguage = doc["language"].as<String>();

  Serial.println();
  Serial.print("Status: ");
  Serial.println(responStatus);
  
  Serial.print("Keterangan: ");
  Serial.println(responKeterangan);

  Serial.print("Language: ");
  Serial.println(responLanguage);
  
  Serial.println("-------------------");
  Serial.println();

  tone(Buzzer, NOTE_E7, 100);
  if (millis() - currentMillis < 120) {
    currentMillis = millis();
    tone(Buzzer, NOTE_E7, 100);
  }

  if (responStatus != "success") {
    lcd.clear();
    lcd.print("Please register");
    lcd.setCursor(2, 1);
    lcd.print("an account.");
    return;
  }

  lcd.clear();
  lcd.print("Success!");
  lcd.setCursor(0, 1);
  lcd.print("Calling " + responNama + ".");

  String ttsString;

  if (responLanguage == "en") {
    ttsString = "This is the pickup call for " + responNama + ". Please proceed to: Gate 1, immediately. Thank you.";
  } else if (responLanguage == "zh" ) {
    ttsString = "这是接听电话 " + responNama + ". 请立即前往：1号登机口。 谢谢。";
  } else if (responLanguage == "ja") {
    ttsString = responNama + "のピックアップコールです" + ". すぐにゲート1に進んでください。 ありがとうございました。";
  } else if (responLanguage == "de") {
    ttsString = "Dies ist der Abholanruf für " + responNama + ". Bitte fahren Sie sofort mit Gate 1 fort. Vielen Dank.";
  } else if (responLanguage == "nl") {
    ttsString = "Dit is het ophaalgesprek voor " + responNama + ". Ga onmiddellijk door naar: Gate 1. Dank je.";
  } else if (responLanguage == "it") {
    ttsString = "Questa è la chiamata di ripresa per " + responNama + ". Procedi immediatamente a: Gate 1. Grazie.";
  } else if (responLanguage == "ru") {
    ttsString = "Это звонок для " + responNama + ". Пожалуйста, немедленно направляйтесь к выходу №1. Спасибо.";
  } else if (responLanguage == "id") {
    ttsString = "Ini adalah panggilan penjemputan untuk " + responNama + ". Silakan keluar melalui: Gerbang 1, segera. Terima kasih.";
  } else if (responLanguage == "jv") {
    ttsString = responNama + "saiki mulih!" + ". Mangga metu liwat: Gerbang 1. Matur suwun.";
  }

  Serial.println(ttsString);
  
  lcd.setCursor(0, 0);
  lcd.print("Please wait.....");
  
  String urlSpeech = tts.getSpeechUrl(ttsString, responLanguage);
  urlSpeech.remove(4, 1);

  Serial.println(urlSpeech.c_str());

  speak(urlSpeech.c_str());
  
  SPI.begin();
  rfid.init();
  Wire.begin(D2, D1);

  lcd.clear();

  Serial.println("Scan another card.");
  
}
```

Any help is appreciated!

**Comment by cyberelectronics** [comment]
I have exactly the same problem.
Two weeks ago worked like a charm, without issues.
Tried on both ESP32 and ESP8266 (latest 2.7.4).
Now (few days ago) it works only 1 to 4 times / 5 minutes, than the ESP hangs, without any error or wdt reboots.
I need to wait again 3 to 5 minutes and the TTS will work again, but only for 1>4 times, if you don't wait a few minutes, will not work at all.
Tried different devices with different MAC/Hostname, using different ISP/mobile hotspot, same issue.
Is it possible that we reached a quota limit from Google side?
Do we need to use/change any "apikey"? 
Here is a simple test sketch:
```
//#include <WiFi.h>      // for ESP32 >> boards manager URL https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json
#include <ESP8266WiFi.h> // for ESP8266, latest 2.7.4
#include <esp8266-google-home-notifier.h>

String  txtnotify;
int counter;
const char* ssid     = "SSID";
const char* password = "Password";
TTS tts;
GoogleHomeNotifier ghn;

void setup() 
{
  Serial.begin(115200);
  Serial.println("");
  Serial.print("connecting to Wi-Fi");
  WiFi.mode(WIFI_STA);
  WiFi.begin(ssid, password);

  while (WiFi.status() != WL_CONNECTED) 
  {
    delay(250);
    Serial.print(".");
  }
  Serial.println("");
  Serial.println("connected.");
  Serial.print("IP address: ");
  Serial.println(WiFi.localIP()); 
  
  const char displayName[] = "Bedroom speaker";    // use CastReceiver android app or Google Home Device Name

  Serial.println("connecting to Google Home...");
  if (ghn.device(displayName, "en") != true) 
  {
    Serial.println(ghn.getLastError());
    return;
  }
  Serial.print("found Google Home(");
  Serial.print(ghn.getIPAddress());
  Serial.print(":");
  Serial.print(ghn.getPort());
  Serial.println(")");

  //ghn.notify("Hello world");

  //Serial.println("Done.");
}

void loop() 
{
  
  counter++;
  txtnotify = "Test ";
  txtnotify = txtnotify + String(counter);
  txtnotify = txtnotify + " ok. ";
  Serial.print(txtnotify);
  ghn.notify(txtnotify.c_str());
  //Serial.println(tts.getSpeechUrl("Hello, World!", "Hi man"));
  Serial.println(" Done");
  
  delay(20000);
  
}
```

**Comment by horihiro** [comment]
Thank you for your feedback.

This looks an issue related to Google translate page also.
It might take a time to find solution for this.

**Comment by timothydillan** [comment]
Hey @horihiro, thanks for your quick response.

I still didn't find a solution to the problem of the library not generating a URL for the second time, but I did find an alternative on generating Google Translate Text-to-speech links, which solved my problem.

While googling around,  I found out that data such as a newly generated token, the length of the message, XL and lltoa is not needed to generate a downloadable and usable TTS link.

Instead, the only things that are needed are the specified language code and the message.

I modified your library so that the only thing left is only the ```urlencode``` and ```getspeechurl``` function.

Here's the new code:
```
String GoogleTTS::getSpeechUrl(String text, String lang) {
    return String("https://translate.google.com/translate_tts?ie=UTF-8&client=tw-ob&tl="+lang+"&q="+this->urlencode(text));
}
```

**Comment by horihiro** [comment]
In short investigation, there are following two interfaces on Google Translate page ( https://translate.google.com/ ) currently.
This library accesses single URL and cannot control to access which interface.

The first one is:
  - This library uses
  - Getting mp3 file **URL**
  - `TKK` parameter **was** needed previously.
    As @timothydillan wrote, it's not needed currently
    ex) https://translate.google.com/translate_tts?ie=UTF-8&client=something&tl=en&q=Hello

The second one is:
  - New interface maybe. 
  - Getting mp3 file encoded as Base64

**Comment by horihiro** [comment]
This is same issue as https://github.com/zlargon/google-tts/issues/33

**Comment by cyberelectronics** [comment]
Just tested @timothydillan solution, with the above test sketch and it works great + it looks like it's much quicker.
Is this a temporary solution?
This video was made with the original library, which is a little bit slower and needed a lot of reboots (until GA TTS responded to me), but hopefully now all issues will be solved.
Thank you for your help! 
https://youtu.be/mUl9IHsN-ao

**Comment by horihiro** [comment]
@timothydillan 's solution is definitely faster than original because getting `TKK` parameter is not needed.
But I'm not sure if the solution can be used for a long time or temporary, because Google changed TTS mechanism on Google Translate page, which is depended on by this library, in many times...

By the way, thank you for sharing the video. it is a great application!

**Comment by cyberelectronics** [comment]
Thank you @horihiro !
This will be opensource, so hopefully anybody can modify, integrate in any car or other projects.
